### PR TITLE
Helm: Add revision history limit for master replica

### PR DESCRIPTION
**Pull request by omerap12**
Fixes: #1759 

Tested locally. Output:
```yaml
---
# Source: node-feature-discovery/templates/master.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name:  release-name-node-feature-discovery-master
  namespace: default
  labels:
    helm.sh/chart: node-feature-discovery-0.2.1
    app.kubernetes.io/name: node-feature-discovery
    app.kubernetes.io/instance: release-name
    app.kubernetes.io/version: "master"
    app.kubernetes.io/managed-by: Helm
    role: master
spec:
  replicas: 1
  revisionHistoryLimit: 3
  selector:
    matchLabels:
      app.kubernetes.io/name: node-feature-discovery
      app.kubernetes.io/instance: release-name
      role: master
  template:
    metadata:
      labels:
        app.kubernetes.io/name: node-feature-discovery
        app.kubernetes.io/instance: release-name
        role: master
    spec:
      serviceAccountName: release-name-node-feature-discovery
      enableServiceLinks: false
      securityContext:
        {}
      containers:
        - name: master
          securityContext:
            allowPrivilegeEscalation: false
            capabilities:
              drop:
              - ALL
            readOnlyRootFilesystem: true
            runAsNonRoot: true
          image: "gcr.io/k8s-staging-nfd/node-feature-discovery:master"
          imagePullPolicy: Always
          livenessProbe:
            grpc:
              port: 8082
            initialDelaySeconds: 10
            periodSeconds: 10
          readinessProbe:
            grpc:
              port: 8082
            initialDelaySeconds: 5
            periodSeconds: 10
            failureThreshold: 10
          ports:
          - containerPort: 8080
            name: grpc
          - containerPort: 8081
            name: metrics
          env:
          - name: NODE_NAME
            valueFrom:
              fieldRef:
                fieldPath: spec.nodeName
          command:
            - "nfd-master"
          resources:
            limits:
              memory: 4Gi
            requests:
              cpu: 100m
              memory: 128Mi
          args:
            ## By default, disable crd controller for other than the default instances
            - "-crd-controller=true"
            # Go over featureGates and add the feature-gate flag
            - "-feature-gates=NodeFeatureAPI=true"
            - "-feature-gates=NodeFeatureGroupAPI=false"
            - "-metrics=8081"
          volumeMounts:
            - name: nfd-master-conf
              mountPath: "/etc/kubernetes/node-feature-discovery"
              readOnly: true
      volumes:
        - name: nfd-master-conf
          configMap:
            name: release-name-node-feature-discovery-master-conf
            items:
              - key: nfd-master.conf
                path: nfd-master.conf
      affinity:
        nodeAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - preference:
              matchExpressions:
              - key: node-role.kubernetes.io/master
                operator: In
                values:
                - ""
            weight: 1
          - preference:
              matchExpressions:
              - key: node-role.kubernetes.io/control-plane
                operator: In
                values:
                - ""
            weight: 1
      tolerations:
        - effect: NoSchedule
          key: node-role.kubernetes.io/master
          operator: Equal
          value: ""
        - effect: NoSchedule
          key: node-role.kubernetes.io/control-plane
          operator: Equal
          value: ""
```

### CODE REVIEW

**Review by marquiz — COMMENTED** [review]
Thanks @omerap12 for the contribution!

For the sake of consistency, I think we should add this parameter to nfd-gc, too (`templates/nfd-gc.yaml`, `gc.revisionHistoryLimit` in values). Also, we need to document the new parameter(s) in `docs/deployment/helm.md`.

File: deployment/helm/node-feature-discovery/values.yaml [hunk 1]
```diff
@@ -90,6 +90,8 @@ master:
     # The name of the service account to use.
     # If not set and create is true, a name is generated using the fullname template
     name:
+
```
Comment (TessaIO) [inline deployment/helm/node-feature-discovery/values.yaml:93]: Probably also a comment here on what this field signifies can be a good addition.

File: deployment/helm/node-feature-discovery/values.yaml [hunk 2]
```diff
@@ -90,6 +90,9 @@ master:
     # The name of the service account to use.
     # If not set and create is true, a name is generated using the fullname template
     name:
+  
+  # specify how many old ReplicaSets for the Deployment to retain.
+  revisionHistoryLimit: 3
```
Comment (TessaIO) [inline deployment/helm/node-feature-discovery/values.yaml:95]: I think you should add the field description in the file located under `docs/deployment/helm.md`

Comment (omerap12) [reply deployment/helm/node-feature-discovery/values.yaml:95]: oh I didn't add the docs lol .

File: docs/deployment/helm.md [hunk 3]
```diff
@@ -144,6 +144,7 @@ API's you need to install the prometheus operator in your cluster.
 | `master.nfdApiParallelism` | integer | 10                                      | Specifies the maximum number of concurrent node updates. |
 | `master.config`            | dict    |                                         | NFD master [configuration](../reference/master-configuration-reference) |
 | `master.args`              | array   | []                                      | Additional [command line arguments](../reference/master-commandline-reference.md) to pass to nfd-master |
+| `master.revisionHistoryLimit` | interger | 3 | Specify how many old ReplicaSets for this Deployment you want to retain. [revisionHistoryLimit](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#revision-history-limit) |
```
Comment (TessaIO) [inline docs/deployment/helm.md:147]: nit
```suggestion
| `master.revisionHistoryLimit` | integer | 3 | Specify how many old ReplicaSets for this Deployment you want to retain. [revisionHistoryLimit](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#revision-history-limit) |
```

File: docs/deployment/helm.md [hunk 4]
```diff
@@ -220,6 +221,7 @@ API's you need to install the prometheus operator in your cluster.
 | `gc.deploymentAnnotations`            | dict   | {}      | Garbage collector deployment [annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/)          |
 | `gc.affinity`                         | dict   | {}      | Garbage collector pod [affinity](https://kubernetes.io/docs/tasks/configure-pod-container/assign-pods-nodes-using-node-affinity/)    |
 | `gc.args`                             | array   | []     | Additional [command line arguments](../reference/gc-commandline-reference.md) to pass to nfd-gc |
+| `gc.revisionHistoryLimit` | interger | 3 | Specify how many old ReplicaSets for this Deployment you want to retain. [revisionHistoryLimit](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#revision-history-limit) |
```
Comment (TessaIO) [inline docs/deployment/helm.md:224]: same here
```suggestion
| `gc.revisionHistoryLimit` | integer | 3 | Specify how many old ReplicaSets for this Deployment you want to retain. [revisionHistoryLimit](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#revision-history-limit) |
```

Comment (TessaIO) [reply docs/deployment/helm.md:224]: also please make sure to fix formatting of the table

File: deployment/helm/node-feature-discovery/values.yaml (hunk 2 above)
Comment (marquiz) [inline deployment/helm/node-feature-discovery/values.yaml:95]: I'd be inclined to use Kubernetes defaults (i.e. leave this empty). Or then have good argumentation why certain value (in this case 3) was picked. WDYT?

Comment (omerap12) [reply deployment/helm/node-feature-discovery/values.yaml:95]: Well the issue asked for setting the value to 3, but I think the default in Kubernetes is 10. I'm good with both :)

**Review by TessaIO — APPROVED** [review]

File: deployment/helm/node-feature-discovery/values.yaml (hunk 2 above)
Comment (marquiz) [reply deployment/helm/node-feature-discovery/values.yaml:95]: Ah, yes. @omerap12 just leave the fields empty (in values.yaml) and we should be good

Comment (omerap12) [reply deployment/helm/node-feature-discovery/values.yaml]: you got it @marquiz

File: docs/deployment/helm.md [hunk 5]
```diff
@@ -144,6 +144,7 @@ API's you need to install the prometheus operator in your cluster.
 | `master.nfdApiParallelism` | integer | 10                                      | Specifies the maximum number of concurrent node updates. |
 | `master.config`            | dict    |                                         | NFD master [configuration](../reference/master-configuration-reference) |
 | `master.args`              | array   | []                                      | Additional [command line arguments](../reference/master-commandline-reference.md) to pass to nfd-master |
+| `master.revisionHistoryLimit` | integer | 3                                    | Specify how many old ReplicaSets for this Deployment you want to retain. [revisionHistoryLimit](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#revision-history-limit) |
```
Comment (marquiz) [inline docs/deployment/helm.md:147]: You need to alter the default in the docs, too :)

Comment (omerap12) [reply docs/deployment/helm.md:147]: LOL

File: docs/deployment/helm.md [hunk 6]
```diff
@@ -217,9 +218,10 @@ API's you need to install the prometheus operator in your cluster.
...
-| `gc.deploymentAnnotations`            | dict   | {}      | Garbage collector deployment [annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/)          |
+| `gc.deploymentAnnotations`            | dict   | {}      | Garbage collector deployment [annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/)           |
 | `gc.affinity`                         | dict   | {}      | Garbage collector pod [affinity](https://kubernetes.io/docs/tasks/configure-pod-container/assign-pods-nodes-using-node-affinity/)    |
-| `gc.args`                             | array   | []     | Additional [command line arguments](../reference/gc-commandline-reference.md) to pass to nfd-gc |
+| `gc.args`                             | array  | []      | Additional [command line arguments](../reference/gc-commandline-reference.md) to pass to nfd-gc |
+| `gc.revisionHistoryLimit`             | integer | 10     | Specify how many old ReplicaSets for this Deployment you want to retain. [revisionHistoryLimit](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#revision-history-limit) |
```
Comment (marquiz) [inline docs/deployment/helm.md:224]: nit: I'd leave this empty as that's (null) the default value we have 

```suggestion
| `gc.revisionHistoryLimit`             | integer |        | Specify how many old ReplicaSets for this Deployment you want to retain. [revisionHistoryLimit](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#revision-history-limit) |
```

**Comment by omerap12** [comment]
Friendly reminder :) @marquiz

**Review by marquiz — APPROVED** [review]
Looks good, now 👍

**Comment by marquiz** [comment]
@ArangoGutierrez @TessaIO maybe we should consider cherry-picking this to v0.16.3. It's such a tiny change even if not strictly a bug fix, WDYT?

**Comment by TessaIO** [comment]
no objections, we can go with that

**Comment by rouke-broersma** [comment]
@omerap12 is there a specific reason you didn't add the field for the daemonset? I think the same applies, no?

**Comment by rouke-broersma** [comment]
Already done :)

**Pull request merged**